Objects/object.c

```diff
@@ -2129,3 +2129,75 @@ _Py_Dealloc(PyObject *op)
 #ifdef __cplusplus
 }
 #endif
+
+
+static struct {
+    PyObject ***array;
+    Py_ssize_t len;
+    Py_ssize_t alloc;
+} once_variables = {NULL, 0, 0};
+
+int
+_PyOnceVar_Set(PyObject **var, PyObject *value)
+{
+    assert(var != NULL);
+    assert(*var == NULL);
```

Comment:
While I agree that the case can occur in practice, I'm not confortable with supporting "reentrant calls" (reentrant calls to _PYONCEVAR_INIT() to be exact). I prefer to start with an assertion, and decide how to handle this corner case if it occurs in practice.

*If* reentrant calls occurs, to be clear: it would already be a bug in CPython, right? So my code doesn't make CPython worse. It's better, since the assertion now detects the bug.